File: 30_days_of_code/15-linked-list/py/linked-list.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None
class Solution:
    def display(self,head):
        current = head
        while current:
            print current.data,
            current = current.next

    def insert(self,head,data):
        if head == None:
            return Node(data)
        else:
            tail_node = self.tail(head)
            tail_node.next = Node(data)
            return head

    def tail(self, head):
        cursor = head
        while cursor.next:
            cursor = cursor.next
        return cursor

mylist= Solution()
T=int(input())
head=None
for i in range(T):
    data=int(input())
    head=mylist.insert(head, data)
mylist.display(head);

File: 30_days_of_code/15-linked-list/cpp/linked-list.cpp
#include <iostream>
using namespace std;

class Node {
public:
  int data;
  Node *next;
  Node(int d) {
    data=d;
    next=NULL;
  }
};

class Solution {
public:
  Node* insert(Node *head, int data)
  {
    if(head == NULL)
      return new Node(data);
    else {
      Node *tail = this->tail(head);
      tail->next = new Node(data);
      return head;
    }
  }

  Node* tail(Node *head)
  {
    Node *cursor = head;
    while(cursor->next)
      cursor = cursor->next;
    return cursor;
  }

  void display(Node *head)
  {
    Node *start=head;
    while(start) {
      cout<<start->data<<" ";
      start=start->next;
    }
  }
};

int main()
{
  Node* head=NULL;
  Solution mylist;
  int T,data;
  cin>>T;
  while(T-->0) {
    cin>>data;
    head=mylist.insert(head,data);
  }
  mylist.display(head);
}
